udev rules: sanitize persistent-storage.rules, and get rid of scsi-id

mwilck opened the pull request:
After discussion with @fbuihuu and @hreinecke, I attempted a rewrite of the udev rules for persistent storage, with the following goals:

 1) get rid of scsi_id, replace by sg calls / sysfs everywhere
    (this implies changes in 55-sg3-utils.rules as well)
 2) cleanly separate hardware detection, content detection (blkid), and
    symlink creation
 3) Improve general readability. For my taste, that means shorter
    lines, thus less conditions individual rules, thus more GOTOs
    separating blocks of code where certain conditins are known to
    hold.

This pull request needs to go together with the corresponding request for sg3_utils: https://github.com/hreinecke/sg3_utils/pull/22

yuwata commented on rules/meson.build:
```diff
@@ -16,6 +16,8 @@
 # along with systemd; If not, see <http://www.gnu.org/licenses/>.
 
 rules = files('''
+        55-storage-detection.rules
```
55-strage-hardware.rules

yuwata commented on test/test-functions:
```diff
@@ -314,7 +314,8 @@ install_dmevent() {
         # dmsetup installs 55-dm and 60-persistent-storage-dm on Debian/Ubuntu
         # see https://anonscm.debian.org/cgit/pkg-lvm/lvm2.git/tree/debian/patches/0007-udev.patch
-        inst_rules 55-dm.rules 60-persistent-storage-dm.rules
+        inst_rules 55-dm.rules 55-storage-hardware-rules \
```
wrong file name again...

mwilck replied:
upps big sorry. I really tested this, no idea how that slipped in.

mwilck commented:
@yuwata, thanks for your thorough review. I fixed the issues you mentioned.

yuwata commented:
Hmm... I think removing `scsi_id` is drastic change. It may be used somewhere. Rather than completely 'kill' it, I prefer that `scsi_id` outputs a deprecate message and announce that somewhere, e.g. NEWS.

yuwata commented:
I am not sure that it can be comparable, but `udevadm hwdb` command is deprecated few years ago, but it still left there for compatibility.

hreinecke commented:
I was about to comment about the same thing; my idea for this whole conversion is to move all calls to scsi_id out of 60-persistent-storage.rules (like Martin did), but keep it in a rule set which complements the sg3_utils rules. The overall idea would be that one could keep one or other (or both or, of course, none); we would then have a series like

55-scsi-sg3_id.rules
58-scsi-sg3_symlink.rules
59-scsi-scsi_id.rules
60-persistent-storage.rules

And one could leave out the first two (if eg sg3_utils is not installed), or the 59 rules (if and when scsi_id gets deprecated), and still get the same user experience.

mwilck commented:
@yuwata, I had the impression that removing scsi_id was a set goal. One reason was the line in the TODO file that my patch removes, another remarks from @poettering I'd seen (e.g. his first comment on https://github.com/systemd/systemd/pull/7190), as well as various discussions with @hreinecke and @fbuihuu. I'm also quite positive that there are hardly any users of scsi_id outside udev/systemd.

However, if you prefer to keep it around, I've no problem with dropping the last patch in the series.

mwilck commented:
I'm sick today, I'll look into the build/packaging failures in detail later.
Obviously the additional rules files may cause packaging errors. They will also require changes in dracut.

fbuihuu commented:
OTOH scsi_id lives in /usr/lib/udev so it's not supposed to be used by anything but udev.

poettering commented:
dropping scsi_id makes a lot of sense to me, but i figure @kaysievers should review this before we merge this.

mwilck commented:
@hreinecke, I'm not sure I understand your proposal correctly. It's crucial to separate 

 1. determination of device properties and creation of udev environment variables
 2. symlink creation based on environment variables
 3. further processing based on environment variables

After applying my patch set, the job of `60-persistent-storage.rules` is 2.) exclusively - creation of symlinks based on certain generic environment variables which all follow the `ID_xxx` naming convention. We don't care any more by which utility these `ID_xxx` variables had been set before. It could be **scsi_id**, **sg_inq**, or whatever else. That's a big step forward.

This distinction should be mapped to rule numbers, too. In particular, **60** should be the separator between 1.) and 2.) above. IOW, SYMLINK calls should start only after 60. The rationale is simple - *be sure all variables are correctly set before starting to set  symlinks* (there are a few exceptions for device-mapper rules currently, which we may want to address at a later stage). 

It's also important to split step 1) above into hardware property detection an content-based property detection, mostly because we need to make "further processing" decisions between these two steps (think multipath). We need a distinction by number here as well, and as multipath is currently at **56**, I chose that. Unfortunately, number space is becoming a bit tight between 55 and 60.

If we decide to keep `scsi_id` calls for setting `ID_xxx` variables, these rules should be placed **after** `55-scsi-sg3_utils.rules` and **before** `56-multipath.rules`. That's a bit tricky but it can be done.

`58-scsi-sg3_symlink.rules` is a bit special. It creates lots of symlinks which are not normally used by systemd or other tools which aren't specific to SCSI. I wonder if anyone except SUSE is using it.

mwilck commented:
There has been no progress for 4 weeks, any suggestions?

mwilck commented:
Rebased, and modified the patch set as follows:

Instead of instantly removing the calls to `scsi_id`, I move them to a separate, new rule file "`55-zz-scsi_id.rules`". This means the rules using `scsi_id` are still in place. So distributions or users can choose using these or using something else (I would recommend **sg3_utils**). The  `scsi_id` rules are only run if nothing else has set `ENV{ID_SERIAL}`.

The last patch removes the new rules file again, together with `scsi_id` itself. The idea is to give systemd maintainers the chance to wait a bit until killing of `scsi_id`, and open a migration path, by applying this series without the last patch. This also unties the dependency to the `sg3_utils` pull request https://github.com/hreinecke/sg3_utils/pull/22 (as long as `scsi_id` rules are in present, no replacement is necessary).

Furthermore, I merged together the commits affecting the same device classes. I think review is easier this way, as you can see better how code is moved.

The udev rules from this patch set have been tested with various SCSI devices including disks, tapes, and medium changers, both using the `scsi_id` rules (all but last patch applied) and using the `sg3_utils` rules. I also verified that `sg3_utils` rules take precedence if both are installed. More testing is of course always welcome.

mwilck commented:
I think this patch set is now final. Compared to the previous push requests, I
 - __removed the commit that dumps scsi_id__. I'd like to do the separation of HW detection, variable setting, and symlink creation to first, and provide a time window for distributions and users to adapt. The [patch ultimately removing scsi_id](https://github.com/mwilck/systemd-1/commit/55676d2e92282dd0af7ea22299366f0f26b3a73f) can be trivially added later if desired.
 - __moved HW detection to "54-"__. This opens a slot at "55-" for SCSI rules to run. That's important because we have to make sure that `usb_id` and `ata_id` are run _before_ SCSI rules. These tools are usually better at determining device properties than generic SCSI tools. Note that this is also the current order in`60-persistent-storage.rules`. Currently (without this patch set) this order is broken if `55-scsi-sg3_id.rules` is present.

I've created a [test script](https://github.com/mwilck/test-udev-rules) to examine the results of udev rule processing with different rule sets and settings in detail, and used it to eliminate errors from this patch set (among other things, this testing motivated the move of HW detection to "54-").

Summarizing again, 3 new udev rule files are added by this patch set:
 - `54-storage-hardware.rules`: rules for HW detection,
 - `55-zz-scsi_id.rules`: rules for SCSI HW detection using `scsi_id`, for systems that don't have `55-scsi-sg3_id.rules` or anything else installed for this purpose,
 - `59-storage-content.rules` with `blkid` calls, to be run after e.g. multipath rules and before `60-persistent-storage.rules`.

__I'd like to ask for review and merging again.__

hreinecke commented:
Patches look good to me. Go.

mwilck commented:
I pushed one more small change: `54-storage-hardware.rules` has been moved to `53-storage-hardware.rules`. This is to allow external packages to place rules deterministally between HW detection and `55-scsi-sg3_utils.rules`. And I fixed a copy/paste error in the rule for IEEE1394 tapes.

fbuihuu commented:
@poettering how could we make progress here ?

@mwilck put a fair amount of work here, it would be unfortunate to lost it

Thanks in adavance

poettering commented:
> @poettering how could we make progress here ?

This is waiting for @kaysievers to review it

mwilck commented:
> This is waiting for @kaysievers to review it

... and has been in this state since > 2 months. @kaysievers, if you absolutely can't spare time for this, please drop a short note, so that we know where we stand.

mwilck commented:
It's four months now. @poettering, how long are we going to wait for @kaysievers review? Is it possible that he lost interest in this topic?

mwilck commented:
This has been waiting for @kaysievers comments for 7 months now.

**What can I do to get this moving?**

poettering commented:
@mwilck heya, let's get this done now. Could you rebase?

mwilck commented:
> @mwilck heya, let's get this done now. Could you rebase?

@poettering: wow, I'd lost hope on this one. Give me a week or so please.

mwilck commented:
@poettering: rebase done.

The companion request for sg3_utils is now https://github.com/hreinecke/sg3_utils/pull/36
The two PRs are independent. The sg3_utils set is only needed in order to finally remove **scsi_id**.

fbuihuu commented on rules/55-zz-scsi_id.rules:
```diff
@@ -0,0 +1,20 @@
+# This is executed after other rules that may have set ID_BUS already
+# (e.g. USB). Don't overwrite it.
+ENV{ID_BUS}!="?*", ENV{.ID_BUS}=="?*", ENV{ID_BUS}="$env{.ID_BUS}"
+
```
Couldn't the value of `ENV{ID_BUS}` be tested at the beginning of this rule file and skip all calls to scsi_id if it's already set ?

This way you a) don't need an temporary variable ".ID_BUS" (the '.' prefix looks a bit weird BTW) b) don't call scsi_id needlessly if ID_BUS is already initialized.

mwilck replied:
> Couldn't the value of `ENV{ID_BUS}` be tested at the beginning of this rule file and skip all calls to scsi_id if it's already set ?

No. This code has to be executed if and only if `ID_SERIAL` is unset yet (likely because **sg3_utils**) was not installed). Testing `ID_BUS` equivalent to testing `ID_SERIAL`. `scsi_id --export` sets lots of environment variables, including  `ID_SERIAL`.
 
> This way you a) don't need an temporary variable ".ID_BUS" (the '.' prefix looks a bit weird BTW) b) don't call scsi_id needlessly if ID_BUS is already initialized.

Well, it is the syntax that **udev** mandates for temporary variables. See **udev(7)**:

>  Property names with a leading "."  are neither stored in the database nor exported to events or external tools (run by, for example, the PROGRAM match key)

fbuihuu replied:
OK, thanks.

fbuihuu commented on rules/55-zz-scsi_id.rules:
```diff
@@ -12,6 +12,15 @@ KERNEL=="sd*[!0-9]|sr*", \
+KERNEL=="st*[0-9]|nst*[0-9]", KERNELS=="[0-9]*:*[0-9]", \
+	ENV{.BSG_DEV}="$root/bsg/$id"
+KERNEL=="st*[0-9]|nst*[0-9]", KERNELS=="[0-9]*:*[0-9]", \
+	IMPORT{program}="scsi_id --whitelisted --export --device=$env{.BSG_DEV}", \
```
Do we really need this `.BSG_DEV` temporary variable ?

mwilck replied:
> Do we really need this `.BSG_DEV` temporary variable ?

Yes, because `IMPORT` doesn't support string substitutions. See **udev(7)** again:

> The NAME, SYMLINK, PROGRAM, OWNER, GROUP, MODE, SECLABEL, and RUN fields support simple string substitutions

For whatever reason, `IMPORT` is not in the list.

fbuihuu replied:
Ok but then what's the point to split this into 2 separate rules ?

mwilck replied:
I believe I tested merging the two rules into one and it didn't work. IIRC it turned out that the environment variable set with `ENV{.BSG_DEV}` was only available in the environment in the following rule, not while the rule itself was executed. The man page doesn't say anything about this. The way it is now, we're on the safe side, perhaps with a tiny performance cost.

fbuihuu replied:
This comment was more about readability than performance cost.

poettering commented:
hmm, i wonder if we shouldn't "pack" the prefix numbers a bit closer. After all they kinda became API, and people might have ordered something before 60- and now things are reordered...

poettering commented on rules/55-zz-scsi_id.rules:
```diff
@@ -0,0 +1,20 @@
+ACTION=="remove", GOTO="scsi_id_end"
```
the naming of this file is strange, none of our other rules files uses underscores. 

maybe call this 56-fallback-scsi-id.rules? that way you don't need the weird "zz" thing, as it is still ordered before multipath anyway

mwilck replied:
My idea was to give some "meaning" to the numbers. 55 for hardware property detection, 58 for content-related property detection, 56 and 57 for stuff that might go between. Using "56-fallback-scsi-id.rules" would violate that principle. "55-zz" looks weird at first sight, but it expresses quite well the intention of the file - being the last thing to execute before "56".
Sticking to the numbers might also make the intention more obvious to future generations of programmers.

But if you insist I can change this, no problem. After all, this is designed to be temporary (famous last words).

poettering replied:
yes, please, the "zz" thing would be new, and if it's temporary we really shouldn't adopt a new naming scheme for that, in particular given that temporary things tend to remain temporary for a long long time

poettering commented:
the numbering feels so terribly fragile...

quite frankly i really don't oversee the implications of all the reordering. I really hope you tested this well.

Could you also add a NEWS file entry for this, explaining what changes for rules files, and everybody else who wants to plug some rules files in here?

mwilck commented:
> hmm, i wonder if we shouldn't "pack" the prefix numbers a bit closer. After all they kinda became API, and people might have ordered something before 60- and now things are reordered...

What's your suggestion then? I'd like to rely on alphabetic ordering as little as possible. My idea was to use the space between 50-udev-default.rules and 60-persistent-storage.rules, leaving "gaps" for future additions, as I've learned when I started programming in Commodore Basic.

mwilck commented:
> the numbering feels so terribly fragile...

Yes, but numbers are better than alphabetic ordering. I'm open to start udev over again with a dependency-based scheme as in systemd. But that would be a larger project.

> quite frankly i really don't oversee the implications of all the reordering. I really hope you tested this well.

I tested this quite a bit, yes. Of course I can't test every device out there. Would you think a "call for testing" on the systemd ML would make sense?

IMO the code is cleaner now; implications of changes are easier to follow than before, when hardware detection, environment variable setting, content detection, and symlink creation were mixed in a single file.
 
> Could you also add a NEWS file entry for this, explaining what changes for rules files, and everybody else who wants to plug some rules files in here?

Sure.

mwilck commented:
Added a draft NEWS entry. Please review. It could be a bit too verbose.

poettering commented:
Looks OK, could you please rebase, and dro the "zz" thing. Should be good to merge then, even for 242, if you hurry. Will set milestone for that for now

poettering commented:
(btw, @mwilck would be great if you could adopt #7409 too maybe if it makes sense, and incorporate this into this new scheme, maybe as a separate follow-up PR though)

poettering commented:
needs a rebase, too now.

poettering commented:
@mwilck ping?

mwilck commented:
@poettering , thanks for the reminder. I haven't forgotten, but unfortunately I'm currently busy with other high prio tasks.

poettering commented:
ping!

mwilck commented:
Rebased and renamed `55-zz-scsi_id.rules` -> `56-fallback-scsi_id.rules` as reqested. Pushing for review, testing TBD. https://github.com/systemd/systemd/pull/7409 requires more attention, I'd like to handle that separately if possible.

mrc0mmand commented:
Few seconds before the timeout:

![image](https://user-images.githubusercontent.com/679338/72738310-968fc780-3ba1-11ea-8ca4-ce0cb4cc9303.png)

Not sure how helpful it's going to be.

mwilck commented:
@mrc0mmand, am I understanding correctly that this patch set is causing your regression?

mwilck commented:
@mrc0mmand, this patch set adds additional udev rules files which dracut doesn't know yet (`53-storage-hardware.rules`, `56-fallback-scsi_id.rules`, `59-storage-content.rules`). Please double check if the initrd generated by dracut includes these files. If it doesn't, failure is expected. You can inject them in the initrd with dracut's `--install` option.

mrc0mmand commented:
> @mrc0mmand, this patch set adds additional udev rules files which dracut doesn't know yet (`53-storage-hardware.rules`, `56-fallback-scsi_id.rules`, `59-storage-content.rules`). Please double check if the initrd generated by dracut includes these files. If it doesn't, failure is expected. You can inject them in the initrd with dracut's `--install` option.

Ah, thank you, that indeed did the trick! For some reason I thought dracut installs all udev rules by default. I'll prepare a patch for CentOS CI shortly.

poettering commented:
ah I missed that this PR is now updated. But requires rebase now.

poettering commented on rules.d/53-storage-hardware.rules:
```diff
@@ -78,4 +78,14 @@ SUBSYSTEM=="ubi", KERNEL=="ubi*_*", ATTRS{mtd_num}=="*", ENV{ID_UBI_MTD}="%s{mtd
+
+ATTRS{ieee1394_id}=="?*", ENV{ID_SERIAL}="$attr{ENV{ID_SERIAL}", ENV{ID_BUS}="ieee1394"
+SUBSYSTEMS=="usb", ATTRS{serial}=="?*", IMPORT{builtin}="usb_id"
+IMPORT{builtin}="path_id"
```
shouldn#t this be split out into a separate file? 54-storage-hardware-tapes.rules, too?

poettering commented:
hmm, if #7409 is desirable, can you maybe take that commit and add it to this PR too, on top of the other stuff? i figure it would need rebasing on top of this change anyway.

alexz-kh commented:
So, 5+ years, issue still appears on Supermicro X10DDW-iN +  Intel® C612 chipset.
any change to catch it ?

mwilck commented:
> So, 5+ years, issue still appears on Supermicro X10DDW-iN + Intel® C612 chipset. any chance to catch it ?

What issue are you referring to concretely?

alexz-kh commented:
@mwilck referenced one https://github.com/systemd/systemd/issues/7157 
tl;tr : same ID_PATH reported by udev

mwilck commented:
see https://github.com/systemd/systemd/issues/7157#issuecomment-1607457833

alexz-kh commented:
@mwilck missed by me! thanks. will reply to origin

mauri commented:
@mwilck Any chance to see this merged in the near future? I see this issue over and over in Ubuntu where sg3_utils is present and that breaks 60-persistent-storage rules, preventing `ata_id` properties to be imported. This have some nasty implications around disks APM capabilities not detected downstream.
